### DISCUSSION TITLE
🤔 패키지 구조: 계층형 vs 도메인형 어떤 것을 선택할까요?

### DISCUSSION THREAD

**Discussion by minju412** [petit-a-petit/MIML#5] [General] [answered]
# 🔸 계층형 디렉터리 구조
스프링 웹 계층의 대표 클래스 혹은 디렉터리들을 기반으로 패키징합니다.

## 예시
```
com
 ㄴ example
     ㄴ server
         ㄴ controller
         ㄴ model
         ㄴ repository
         ㄴ service
```

## 장단점
장점
- 전체 구조를 빠르게 파악할 수 있다.

단점
-  각각의 패키지 디렉터리에 클래스들이 너무 많이 모이게 된다.

# 🔸 도메인형 디렉터리 구조
스프링 웹 계층에 주목하기 보다는 도메인에 주목합니다.

## 예시
```
com
 ㄴ example
     ㄴ server
         ㄴ domain
         |   ㄴ member
         |   |   ㄴ controller
         |   |   ㄴ service
         |   |   ㄴ model
         |   |   ㄴ repository
         |   ㄴ pet
         |   |   ㄴ controller
         |   |   ㄴ service
         |   |   ㄴ model
         |   |   ㄴ repository
         |   ...
         ㄴ global
             ㄴ auth
             ㄴ config
             ㄴ exception
             ㄴ util
```

## 장단점
장점
- 각각의 도메인 별로 패키지 분리가 가능하여 관리에 있어서 계층형 방식보다 직관적이다.
- 각각의 도메인들이 서로를 의존하는 코드가 없도록 설계하기 적합하다.

단점
- 도메인의 수가 많아질 수록 패키지의 수가 많아지기 때문에 전체 구조를 파악하기 어려울 수 있다.
- 개발자의 성향에 따라 도메인을 구분하는 기준이 달라, 내 예상과 다른 패키지에 존재하는 경우 관련 코드를 찾기 어렵다.

# 🔸 개인적인 생각
네 명이서 동시에 개발하는 만큼 도메인이 서로 겹치지 않는게 중요할 것 같아요.
이러한 점에서 각각의 도메인별로 의존하는 코드가 없도록 설계할 수 있는 "도메인형 디렉터리 구조"를 사용하면 어떨까 싶습니다!

각각 장단점이 있기 때문에 다양한 의견 주시면 좋을 것 같아요 :)

**Comment by Nine-JH** [comment, marked as answer]
저는 도메인이 좋은 것 같아요. 추후 프로젝트 분리를 할 때에도 직관적일 것 같아요 😎
